### PR TITLE
implement new ComponentArray and change existing code to use it

### PR DESCRIPTION
**Pull request by mechaffin**
# Type of change? 

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

Implementation for https://github.com/Azure/pcs-remote-monitoring-webui/issues/1105

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/pcs-remote-monitoring-webui/1108)

### CODE REVIEW

File: src/components/pages/packages/packages.js [hunk 1]
```diff
@@ -62,25 +63,27 @@ export class Packages extends Component {
...
-        <PageTitle className="package-title" titleValue={t('packages.title')} />
-        {!!error && <AjaxError t={t} error={error} />}
-        {!error && <PackagesGrid {...gridProps} />}
-        {this.state.openFlyoutName === 'new-Package' && <PackageNewContainer t={t} onClose={this.closeFlyout} />}
-      </PageContent>
-    ];
+    return (
+      <ComponentArray>
+        <ContextMenu>
+          <ContextMenuAlign key="left" left={true}>
```
Comment (spryor) [inline src/components/pages/packages/packages.js]: Doesn't seem like you need these `key` attributes

File: src/components/pages/packages/packagesGrid/packagesGrid.js [hunk 2]
```diff
@@ -30,11 +30,12 @@ export class PackagesGrid extends Component {
...
     ];
 
-    this.contextBtns = [
-      <Protected key="delete" permission={permissions.deletePackages}>
-        <Btn svg={svgs.trash} onClick={this.openFlyout('delete-package')}>{props.t('packages.delete')}</Btn>
-      </Protected>
-    ];
+    this.contextBtns =
+      <ComponentArray>
+        <Protected key="delete" permission={permissions.deletePackages}>
```
Comment (spryor) [inline src/components/pages/packages/packagesGrid/packagesGrid.js]: `key` not needed

File: src/components/pages/rules/rules.js [hunk 3]
```diff
@@ -87,29 +88,31 @@ export class Rules extends Component {
...
-        <PageTitle titleValue={t('rules.title')} />
-        { !!error && <AjaxError t={t} error={error} /> }
-        {!error && <RulesGrid {...gridProps} />}
-        {this.state.openFlyoutName === 'newRule' && <NewRuleFlyout t={t} onClose={this.closeFlyout} logEvent={logEvent} />}
-      </PageContent>
-    ];
+    return (
+      <ComponentArray>
+        <ContextMenu>
+          <ContextMenuAlign key="left" left={true}>
```
Comment (spryor) [inline src/components/pages/rules/rules.js]: `key` not needed

File: src/components/pages/rules/rulesGrid/rulesGrid.js [hunk 4]
```diff
@@ -60,21 +60,21 @@ export class RulesGrid extends Component {
           </Btn>
         </Protected>,
       edit:
-        <Protected key="edit"permission={permissions.updateRules}>
+        <Protected key="edit" permission={permissions.updateRules}>
```
Comment (spryor) [inline src/components/pages/rules/rulesGrid/rulesGrid.js]: These might not be needed.

File: src/components/pages/rules/rules.js [hunk 5]
```diff
@@ -87,29 +88,31 @@ export class Rules extends Component {
...
+            <DeviceGroupDropdown />
+            <Protected permission={permissions.updateDeviceGroups}>
+              <ManageDeviceGroupsBtn />
+            </Protected>
+          </ContextMenuAlign>
+          <ContextMenuAlign key="right">
```
Comment (spryor) [inline src/components/pages/rules/rules.js]: `key` not needed

File: src/components/pages/packages/packages.js [hunk 6]
```diff
@@ -62,25 +63,27 @@ export class Packages extends Component {
...
+            { /* Add left aligned items as needed */}
+          </ContextMenuAlign>
+          <ContextMenuAlign key="right">
```
Comment (spryor) [inline src/components/pages/packages/packages.js]: `key` not needed

File: src/components/pages/maintenance/summary/summary.js [hunk 7]
```diff
@@ -38,8 +39,9 @@ export const Summary = ({
   onTimeIntervalChange,
   timeInterval,
   ...props
-}) => [
-    <ContextMenu key="context-menu">
+}) =>
+  <ComponentArray>
+    <ContextMenu>
       <ContextMenuAlign key="left" left={true}>
```
Comment (spryor) [inline src/components/pages/maintenance/summary/summary.js]: `key` not needed

File: src/components/pages/maintenance/ruleDetails/ruleDetails.js [hunk 8]
```diff
@@ -281,138 +283,139 @@ export class RuleDetails extends Component {
...
+                      rowData={isPending ? undefined : this.state.devices}
+                      onContextMenuChange={this.onContextMenuChange('deviceContextBtns')}
+                      onHardSelectChange={this.onHardSelectChange('devices')} />
+                  </ComponentArray>
+                }
+                {
+                  !isPending && (selectedTab === tabIds.all || selectedTab === tabIds.telemetry) && Object.keys(this.state.telemetry).length > 0 &&
+                  <ComponentArray>
+                    <h4 className="sub-heading" key="header">{t('maintenance.alertedDeviceTelemetry')}</h4>
+                    <div className="details-chart-container" key="chart">
```
Comment (spryor) [inline src/components/pages/maintenance/ruleDetails/ruleDetails.js]: `key` not needed here and the line above

File: src/components/pages/maintenance/ruleDetails/ruleDetails.js [hunk 9]
```diff
@@ -281,138 +283,139 @@ export class RuleDetails extends Component {
...
-            || this.state.alertContextBtns
-            || this.state.deviceContextBtns
-          }
-          <TimeIntervalDropdown
-            onChange={onTimeIntervalChange}
-            value={timeInterval}
+    return (
+      <ComponentArray>
+        <ContextMenu className="rule-details-context-menu-container">
+          <ContextMenuAlign key="left" left={true}>
```
Comment (spryor) [inline src/components/pages/maintenance/ruleDetails/ruleDetails.js]: `key` not needed

File: src/components/pages/maintenance/ruleDetails/ruleDetails.js [hunk 10]
```diff
@@ -281,138 +283,139 @@ export class RuleDetails extends Component {
...
+            <DeviceGroupDropdown />
+            <Protected permission={permissions.updateDeviceGroups}>
+              <ManageDeviceGroupsBtn />
+            </Protected>
+          </ContextMenuAlign>
+          <ContextMenuAlign key="right">
```
Comment (spryor) [inline src/components/pages/maintenance/ruleDetails/ruleDetails.js]: `key` not needed

File: src/components/pages/devices/flyouts/deviceDetails/deviceDetails.js [hunk 11]
```diff
@@ -342,18 +344,19 @@ export class DeviceDetails extends Component {
                         <Cell className="col-7">{device.connected ? t('devices.flyouts.details.connected') : t('devices.flyouts.details.notConnected')}</Cell>
                       </Row>
                       {
-                        device.connected && [
+                        device.connected &&
+                        <ComponentArray>
                           <Row key="diag-row-time">
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDetails/deviceDetails.js]: `key` not needed

File: src/components/pages/devices/flyouts/deviceDetails/deviceDetails.js [hunk 12]
```diff
@@ -342,18 +344,19 @@ export class DeviceDetails extends Component {
...
                             <Cell className="col-3">{t('devices.flyouts.details.diagnostics.lastMessage')}</Cell>
-                            <Cell className="col-7">{ lastMessageTime ? moment(lastMessageTime).format(DEFAULT_TIME_FORMAT) : '---' }</Cell>
+                            <Cell className="col-7">{lastMessageTime ? moment(lastMessageTime).format(DEFAULT_TIME_FORMAT) : '---'}</Cell>
                           </Row>,
                           <Row key="diag-row-msg">
```
Comment (spryor) [inline src/components/pages/devices/flyouts/deviceDetails/deviceDetails.js]: `key` not needed

File: src/components/pages/devices/devices.js [hunk 13]
```diff
@@ -76,33 +77,35 @@ export class Devices extends Component {
...
-        {!!error && <AjaxError t={t} error={error} />}
-        {!error && <DevicesGrid {...gridProps} />}
-        {newDeviceFlyoutOpen && <DeviceNewContainer onClose={this.closeFlyout} />}
-        {simManagementFlyoutOpen && <SIMManagementContainer onClose={this.closeFlyout} />}
-      </PageContent>
-    ];
+    return (
+      <ComponentArray>
+        <ContextMenu>
+          <ContextMenuAlign key="left" left={true}>
```
Comment (spryor) [inline src/components/pages/devices/devices.js]: `key` not needed

File: src/components/pages/devices/devices.js [hunk 14]
```diff
@@ -76,33 +77,35 @@ export class Devices extends Component {
...
+            <DeviceGroupDropdown />
+            <Protected permission={permissions.updateDeviceGroups}>
+              <ManageDeviceGroupsBtn />
+            </Protected>
+          </ContextMenuAlign>
+          <ContextMenuAlign key="right">
```
Comment (spryor) [inline src/components/pages/devices/devices.js]: `key` not needed

File: src/components/pages/deployments/flyouts/deploymentNew/deploymentNew.js [hunk 15]
```diff
@@ -269,13 +270,19 @@ export class DeploymentNew extends LinkedComponent {
...
                   deviceFetchSuccessful &&
-                  [
-                    <SummaryCount key="deviceSummaryCount"> {targetedDeviceCount}</SummaryCount>,
-                    <SectionDesc key="deviceSectionDesc">{t('deployments.flyouts.new.targetText')}</SectionDesc>,
-                    completedSuccessfully && <Svg key="deviceSvg" className="summary-icon" path={svgs.apply} />,
-                  ]
+                  <ComponentArray>
+                    <SummaryCount> {targetedDeviceCount}</SummaryCount>
+                    <SectionDesc>{t('deployments.flyouts.new.targetText')}</SectionDesc>
+                    {completedSuccessfully && <Svg key="deviceSvg" className="summary-icon" path={svgs.apply} />}
```
Comment (spryor) [inline src/components/pages/deployments/flyouts/deploymentNew/deploymentNew.js]: `key` not needed

File: src/components/pages/deployments/deploymentsGrid/deploymentsGrid.js [hunk 16]
```diff
@@ -50,8 +50,8 @@ export class DeploymentsGrid extends Component {
       }
     };
 
-    return ([
+    return (
       <PcsGrid {...gridProps} key="deployments-grid" />
```
Comment (spryor) [inline src/components/pages/deployments/deploymentsGrid/deploymentsGrid.js]: `key` not needed

File: src/components/pages/deployments/deployments.js [hunk 17]
```diff
@@ -66,28 +67,30 @@ export class Deployments extends Component {
...
-        <PageTitle className="deployments-title" titleValue={t('deployments.title')} />
-        {!!error && <AjaxError t={t} error={error} />}
-        {!error && <DeploymentsGrid {...gridProps} />}
-        {this.state.openFlyoutName === 'newDeployment' && <DeploymentNewContainer t={t} onClose={this.closeFlyout} />}
-      </PageContent>
-    ];
+    return (
+      <ComponentArray>
+        <ContextMenu>
+          <ContextMenuAlign key="left" left={true}>
```
Comment (spryor) [inline src/components/pages/deployments/deployments.js]: `key` not needed

File: src/components/pages/deployments/deployments.js [hunk 18]
```diff
@@ -66,28 +67,30 @@ export class Deployments extends Component {
...
+            <DeviceGroupDropdown />
+            <Protected permission={permissions.updateDeviceGroups}>
+              <ManageDeviceGroupsBtn />
+            </Protected>
+          </ContextMenuAlign>
+          <ContextMenuAlign key="right">
```
Comment (spryor) [inline src/components/pages/deployments/deployments.js]: `key` not needed

File: src/components/pages/dashboard/dashboard.js [hunk 19]
```diff
@@ -355,90 +356,93 @@ export class Dashboard extends Component {
...
+    return (
+      <ComponentArray>
+        <ContextMenu>
+          <ContextMenuAlign key="left" left={true}>
+            <DeviceGroupDropdown />
+            <Protected permission={permissions.updateDeviceGroups}>
+              <ManageDeviceGroupsBtn />
+            </Protected>
+          </ContextMenuAlign>
+          <ContextMenuAlign key="right">
```
Comment (spryor) [inline src/components/pages/dashboard/dashboard.js]: `key` not needed

File: src/components/pages/dashboard/dashboard.js [hunk 20]
```diff
@@ -355,90 +356,93 @@ export class Dashboard extends Component {
...
-            <OverviewPanel
-              activeDeviceGroup={activeDeviceGroup}
-              openWarningCount={openWarningCount}
-              openCriticalCount={openCriticalCount}
-              onlineDeviceCount={onlineDeviceCount}
-              offlineDeviceCount={offlineDeviceCount}
+    return (
+      <ComponentArray>
+        <ContextMenu>
+          <ContextMenuAlign key="left" left={true}>
```
Comment (spryor) [inline src/components/pages/dashboard/dashboard.js]: `key` not needed

File: src/components/pages/devices/flyouts/deviceDetails/deviceDetails.js (hunk 12 above)
Comment (mechaffin) [reply src/components/pages/devices/flyouts/deviceDetails/deviceDetails.js]: also comma in line above this is wrong too

File: src/components/pages/rules/rulesGrid/rulesGrid.js (hunk 4 above)
Comment (mechaffin) [reply src/components/pages/rules/rulesGrid/rulesGrid.js]: It seems to want these.

**Pull request merged**